### PR TITLE
eglstream-kms: Fully functional platform

### PR DESCRIPTION
**Pull request by RAOF**
Fix up the terrible flickering with the new NVIDIA drivers, and implement a `WaylandAllocator` that returns buffers supporting the new `mg::gl::Texture` interface.

### CODE REVIEW

**Comment by RAOF** [comment]
Oh, yeah. And…
bors try

**Comment by RAOF** [comment]
Now tested on multiple monitors; it *works*, but (at least) clone mode is somewhat flickery - we might be double-rendering or maybe double-acquiring buffers or something.

It appears to work without issue on a single output, and this is strictly better than the status quo, so should be ready for review (once prerequisites land).

**Comment by AlanGriffiths** [comment]
That doesn't work (nor does the nearby similar code selecting the linker). Because:
```
# DEB_DISTRIBUTION: the distribution(s) listed in the current entry of debian/changelog
```
... and that's "UNRELEASED".

**Comment by RAOF** [comment]
This can be built against https://github.com/NVIDIA/egl-wayland , (which will also do nice things like provede `pkg-config` files rather than me listing a bunch of possible library names) but that's not packaged for Ubuntu or Debian yet.

**Comment by RAOF** [comment]
Ooooh! I did not know Fedora had that packaged!

The libnvidia-egl-wayland.so in there *should* contain the necessary symbols. I'll spool up a Fedora container on Monday.

**Comment by Conan-Kudo** [comment]
@RAOF Yeah, apparently we have those libraries packaged, along with the pkg-config stuff. But as @AlanGriffiths pointed out, it doesn't work. :(

**Comment by gerboland** [comment]
This is crashing on me on startup. I'm testing on Bionic, kms is enabled, nvidia driver 390.77-0ubuntu0.18.04.1. Log and stacktrace:

```
Starting program: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/miral-shell
MIR_CLIENT_PLATFORM_PATH=/home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/client-modules/
MIR_SERVER_PLATFORM_PATH=/home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/
LD_LIBRARY_PATH=/home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib
exec=/home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/miral-shell.bin
process 3597 is executing new program: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/miral-shell.bin
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[2018-11-26 12:04:48.852208] mirserver: Starting
[2018-11-26 12:04:48.855483] <DEBUG> mirserver: Using logind for session management
[2018-11-26 12:04:48.855616] mircommon: Loading modules from: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/
[2018-11-26 12:04:48.855695] mircommon: Loading module: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/server-mesa-x11.so.15
[2018-11-26 12:04:48.855723] mircommon: Loading module: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/graphics-mesa-kms.so.15
[2018-11-26 12:04:48.855738] mircommon: Loading module: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/graphics-eglstream-kms.so.15
[2018-11-26 12:04:48.855749] mircommon: Loading module: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/input-evdev.so.7
[2018-11-26 12:04:48.920601] mircommon: Loading module: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/input-stub.so
[2018-11-26 12:04:48.943014] mircommon: Loading module: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/graphics-throw.so
[2018-11-26 12:04:48.943066] mircommon: Loading module: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/graphics-dummy.so
[2018-11-26 12:04:48.943096] mirplatform: Found graphics driver: mir:mesa-x11 (version 1.0.0)
[2018-11-26 12:04:49.108478] mesa-kms: Detected software renderer: llvmpipe (LLVM 6.0, 256 bits)
[2018-11-26 12:04:49.109290] mirplatform: Found graphics driver: mir:mesa-kms (version 1.0.0)
[2018-11-26 12:04:49.109349] <DEBUG> eglstream: Found EGLDeviceEXT with device extensions: EGL_NV_device_cuda EGL_EXT_device_drm
[2018-11-26 12:04:49.163758] mirplatform: Found graphics driver: mir:eglstream-kms (version 1.0.0)
[2018-11-26 12:04:49.163833] mirplatform: Found graphics driver: throw-on-creation (version 1.0.0)
[2018-11-26 12:04:49.163847] mirplatform: Found graphics driver: mir:stub-graphics (version 1.0.0)
[2018-11-26 12:04:49.167666] mirserver: Selected driver: mir:eglstream-kms (version 1.0.0)
[2018-11-26 12:04:49.347625] mirserver: Using software cursor
[2018-11-26 12:04:49.348786] mircommon: Loading modules from: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/
[2018-11-26 12:04:49.348843] mircommon: Loading module: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/server-mesa-x11.so.15
[2018-11-26 12:04:49.348858] mircommon: Loading module: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/graphics-mesa-kms.so.15
[2018-11-26 12:04:49.348878] mircommon: Loading module: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/graphics-eglstream-kms.so.15
[2018-11-26 12:04:49.348897] mircommon: Loading module: /home/gerry/dev/Canonical/build-mir-Desktop-Default/bin/../lib/server-modules/input-evdev.so.7
[2018-11-26 12:04:49.453688] mirserver: Selected input driver: mir:evdev-input (version: 1.0.0)
[2018-11-26 12:04:49.462338] platform-eglstream-kms: Bound EGLStreams-backed WaylandAllocator display

Thread 1 "miral-shell.bin" received signal SIGSEGV, Segmentation fault.
0x00007ffff7e4eec8 in mir::udev::Monitor::fd (this=0x0) at /home/gerry/dev/Canonical/mir/src/platform/udev/udev_wrapper.cpp:329
329         return udev_monitor_get_fd(monitor);
(gdb) bt
#0  0x00007ffff7e4eec8 in mir::udev::Monitor::fd() const (this=0x0) at /home/gerry/dev/Canonical/mir/src/platform/udev/udev_wrapper.cpp:329
#1  0x00007fffef348e7e in mir::graphics::eglstream::Display::register_configuration_change_handler(mir::graphics::EventHandlerRegister&, std::function<void ()> const&) (this=0x55555571e610, handlers=
    ..., conf_change_handler=...) at /home/gerry/dev/Canonical/mir/src/platforms/eglstream-kms/server/display.cpp:405
#2  0x00007ffff5e5cac5 in mir::DisplayServer::Private::Private(mir::ServerConfiguration&) (this=0x55555571c700, config=...) at /home/gerry/dev/Canonical/mir/src/server/display_server.cpp:66
#3  0x00007ffff5e5c197 in mir::DisplayServer::DisplayServer(mir::ServerConfiguration&) (this=0x7fffffffdd68, config=...) at /home/gerry/dev/Canonical/mir/src/server/display_server.cpp:215
#4  0x00007ffff5e56df8 in mir::run_mir(mir::ServerConfiguration&, std::function<void (mir::DisplayServer&)>, std::function<void (int)> const&) (config=..., init=..., terminator=...)
    at /home/gerry/dev/Canonical/mir/src/server/run_mir.cpp:98
#5  0x00007ffff5e8d18e in mir::Server::run() (this=0x5555556e2d40) at /home/gerry/dev/Canonical/mir/src/server/server.cpp:410
#6  0x00007ffff7f0556b in miral::MirRunner::Self::run_with(std::initializer_list<std::function<void (mir::Server&)> >) (this=0x5555556dc910, options=...)
    at /home/gerry/dev/Canonical/mir/src/miral/runner.cpp:209
#7  0x00007ffff7f05b0c in miral::MirRunner::run_with(std::initializer_list<std::function<void (mir::Server&)> >) (this=0x7fffffffdfa0, options=...)
    at /home/gerry/dev/Canonical/mir/src/miral/runner.cpp:256
#8  0x00005555555b4c65 in main(int, char const**) (argc=1, argv=0x7fffffffe468) at /home/gerry/dev/Canonical/mir/examples/miral-shell/shell_main.cpp:95
```

**Comment by AlanGriffiths** [comment]
This builds on Fedora 29 with egl-wayland-devel 1.1.0-0.2.20181015git0eb29d4.fc29 (currently on updates-testing). I've put the necessary changes onto a separate branch (#648)

**Comment by gerboland** [comment]
Latest push fixes my crash on startup. I am trying to test clients with it, but they all die in various ways. glmark2-wayland should work, no? My (messy) log: https://pastebin.ubuntu.com/p/bXdfhQZKmM/

**Comment by gerboland** [comment]
aha ok, I've got glmark2-wayland rendering smoothly with mir_demo_server! Though glmark2-wayland did crash on shutdown.

miral-shell I'm having no luck with, both internal and external clients fail with

```
[2018-11-27 11:40:57.200121] <CRITICAL> frontend:Wayland: Exception processing Surface::commit() request: /home/gerry/dev/Canonical/mir/src/platforms/eglstream-kms/server/buffer_allocator.cpp(288): Throw in function {anonymous}::BoundEGLStream::TextureHandle::TextureHandle(std::shared_ptr<{anonymous}::Sync>, std::shared_ptr<const {anonymous}::EGLStreamTextureConsumer>)
Dynamic exception type: boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::system_error> >
std::exception::what: Failed to latch texture from client EGLStream: EGL_BAD_STATE_KHR (0x321c)
```

Known issue?

**Comment by gerboland** [comment]
I installed nvidia 410 driver from https://launchpad.net/~graphics-drivers/+archive/ubuntu/ppa?field.series_filter=bionic
With that, miral-shell works with glmark2-wayland too, as well as the Qt5 and GTK3/4 apps. miral-shell's internal client not working still. And I tried the demo mir clients, software ones aren't working (not end of the world..)

I do think we need to either support the older nvidia driver, or error out if it is detected. Mir providing a poor experience will end up with Mir getting the blame, even if it is a driver issue.

**Comment by RAOF** [comment]
Not yet tested on my NVIDIA laptop, which needs a little TLC.

This should now detect the version of the NVIDIA driver in probe() and return `unsupported` on NVIDIA drivers earlier than the 396 series (which is what Cosmic has, which works). You can still explicitly load the `eglstream-kms` platform and it'll bypass this check.

**Comment by AlanGriffiths** [comment]
Am I understanding correctly: with this, Mir only supports the Nvidia drivers >= 396, which are available in the Cosmic archive and from ppa:graphics-drivers/ppa for the LTS releases?

How will that affect Mir snaps?

- Should (can) we add ppa:graphics-drivers/ppa (and nvidia-410?) to Intel builds of the mir-kiosk snap?
- Should (can) we add ppa:graphics-drivers/ppa (and nvidia-410?) to Intel builds of the egmde snap?

(I know long term this should be punted to a "platform graphics" interface but, if we want to demo this we need a short term solution.)

**Comment by Conan-Kudo** [comment]
The Fedora `egl-wayland` update has bit stable updates.

**Comment by Conan-Kudo** [comment]
The Fedora `egl-wayland` update has hit stable updates in Fedora.

File: CMakeLists.txt [hunk 1]
```diff
@@ -208,14 +208,26 @@ endif(MIR_DISABLE_EPOLL_REACTOR)
...
 
-# Default to KMS backend, but build all of them
-set(
-  MIR_PLATFORM
-  mesa-kms;mesa-x11;eglstream-kms
-  CACHE
-  STRING
-  "a list of graphics backends to build (options are 'mesa-kms', 'mesa-x11', or 'eglstream-kms')"
-)
+find_library(NVIDIA_EGL_WAYLAND NAMES nvidia-egl-wayland libnvidia-egl-wayland.so.1 libnvidia-egl-wayland.so.1.0.2 libnvidia-egl-wayland.so.1.0.3)
```
Comment (Conan-Kudo) [inline CMakeLists.txt]: Can't we just use the pkgconfig names here? `wayland-eglstream` and `wayland-eglstream-protocols` are the pkgconfig module names.

Comment (AlanGriffiths) [reply CMakeLists.txt]: That would be nice, but AIUI the Ubuntu packages are somewhat behind Fedora.

Of course, we should try pkgconfig first.

**Comment by Saviq** [comment]
>     * Should (can) we add ppa:graphics-drivers/ppa (and nvidia-410?) to Intel builds of the mir-kiosk snap?
>     * Should (can) we add ppa:graphics-drivers/ppa (and nvidia-410?) to Intel builds of the egmde snap?

No, IMO this needs to be solved by error messages and documentation - we can even point at a URL in the error message describing the problem and the solution.

**Comment by gerboland** [comment]
IMO this is worth landing

**Pull request merged**